plugins/crypto/mbedtls/ua_pki_mbedtls.c

```diff
@@ -236,251 +238,158 @@ reloadCertificates(CertInfo *ci) {
+            prev = NULL;
+        }
+        /* Switch to search in the ctx->skIssue list */
+        stack = (stack != &ci->certificateIssuerList) ? &ci->certificateIssuerList : NULL;
+    } while(stack);
+    return NULL;
+}
 
-    if(ci->trustListFolder.length == 0 &&
-       ci->issuerListFolder.length == 0 &&
-       ci->revocationListFolder.length == 0 &&
-       ci->certificateTrustList.raw.len == 0 &&
-       ci->certificateIssuerList.raw.len == 0 &&
-       ci->certificateRevocationList.raw.len == 0) {
-        UA_LOG_WARNING(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
-                       "PKI plugin unconfigured. Accepting the certificate.");
-        return UA_STATUSCODE_GOOD;
+static UA_Boolean
+mbedtlsCheckRevoked(CertInfo *ci, mbedtls_x509_crt *cert) {
+    for(mbedtls_x509_crl *crl = &ci->certificateRevocationList; crl; crl = crl->next) {
+        if(x509_name_eq(&cert->subject, &crl->issuer))
+            return (mbedtls_x509_crt_is_revoked(cert, crl) != 0);
     }
+    return false;
+}
 
-    /* Parse the certificate */
-    mbedtls_x509_crt remoteCertificate;
-
-    /* Temporary Object to parse the trustList */
-    mbedtls_x509_crt *tempCert = NULL;
-
-    /* Temporary Object to parse the revocationList */
-    mbedtls_x509_crl *tempCrl = NULL;
-
-    /* Temporary Object to identify the parent CA when there is no intermediate CA */
-    mbedtls_x509_crt *parentCert = NULL;
-
-    /* Temporary Object to identify the parent CA when there is intermediate CA */
-    mbedtls_x509_crt *parentCert_2 = NULL;
-
-    /* Flag value to identify if the issuer certificate is found */
-    int issuerKnown = 0;
-
-    /* Flag value to identify if the parent certificate found */
-    int parentFound = 0;
+#define UA_MBEDTLS_MAX_CHAIN_LENGTH 10
 
-    mbedtls_x509_crt_init(&remoteCertificate);
-    int mbedErr = mbedtls_x509_crt_parse(&remoteCertificate, certificate->data,
-                                         certificate->length);
-    if(mbedErr) {
-        /* char errBuff[300]; */
-        /* mbedtls_strerror(mbedErr, errBuff, 300); */
-        /* UA_LOG_WARNING(data->policyContext->securityPolicy->logger, UA_LOGCATEGORY_SECURITYPOLICY, */
-        /*                "Could not parse the remote certificate with error: %s", errBuff); */
-        return UA_STATUSCODE_BADSECURITYCHECKSFAILED;
+static UA_StatusCode
+mbedtls_verifyChain(CertInfo *ci, mbedtls_x509_crt *cert,
+                    mbedtls_x509_crt **old_issuers, int depth) {
+    /* Maxiumum chain length */
+    if(depth == UA_MBEDTLS_MAX_CHAIN_LENGTH)
+        return UA_STATUSCODE_BADCERTIFICATECHAININCOMPLETE;
+
+    /* Verification Step: Validity Period */
+    if(mbedtls_x509_time_is_future(&cert->valid_from) ||
+       mbedtls_x509_time_is_past(&cert->valid_to))
+        return UA_STATUSCODE_BADCERTIFICATETIMEINVALID;
+
+    /* Verification Step: Revocation Check */
+    if(mbedtlsCheckRevoked(ci, cert))
+        return UA_STATUSCODE_BADCERTIFICATEREVOKED;
```

Comment:
What happens if no CRL is found here? Shouldn't it return `UA_STATUSCODE_BADCERTIFICATEREVOCATIONUNKNOWN` or `UA_STATUSCODE_BADCERTIFICATEISSUERREVOCATIONUNKNOWN`?

Reply:
The standard says the following:
> Each CA [Certificate](https://reference.opcfoundation.org/search/357?t=Certificate) may have a revocation list. This check fails if this list is not available (i.e. a network interruption prevents the application from accessing the list). No error is reported if the [Administrator](https://reference.opcfoundation.org/search/357?t=Administrator) disables revocation checks for a CA [Certificate](https://reference.opcfoundation.org/search/357?t=Certificate).

To me this is something for certificates with a defined CRL distribution point (e.g. HTTP URL to download the CRL).

We have many "casual users" who just dump a list of trusted certificates into their config.
But who do not have a CRL for all CA or self-signed certificates.

Reply:
The standard is not clear about this point. Reading the description of "AddCertificate()"

> This [Method](https://reference.opcfoundation.org/search/49?t=Method) cannot provide CRLs so issuer [Certificates](https://reference.opcfoundation.org/search/49?t=Certificates) cannot be added with this [Method](https://reference.opcfoundation.org/search/49?t=Method). Instead, CA [Certificates](https://reference.opcfoundation.org/search/49?t=Certificates) and their CRLs shall be managed with the [Write Method](https://reference.opcfoundation.org/search/49?t=Write%20Method) on the containing [TrustList Object.](https://reference.opcfoundation.org/search/49?t=TrustList%20Object.)

we conclude that CRLs are mandatory for all CAs.

Reply:
The trustlist data type looks like this:
![grafik](https://github.com/user-attachments/assets/c1281662-816b-482e-9f61-24e88a187e6c)
And there are a couple of enums to define whether CRLs are provided or not.

I think we should handle the CRL presence check in the GDS code. Not in the underlying crypto plugin.
So we can do it just once for OpenSSL and mbedTLS.
At least for 1.3 and 1.4 we will not make the (breaking) change.

Reply:
I don't known if I can follow. Isn't it required to parse the certificates and CRLs using the underlying crypto library to do the matching? I would assume that most code already have to be in the crypto plugin. So either you end up parsing stuff multiple times our I just do not understand your proposal.